Documentation/architecture.rst

```diff
@@ -0,0 +1,155 @@
+.. only:: not (epub or latex or html)
+
+    WARNING: You are looking at unreleased Cilium documentation.
+    Please use the official rendered version released here:
+    http://docs.cilium.io
+
+.. _arch_guide:
+
+############
+Architecture
+############
+
+This document describes the Cilium architecture. It focuses on
+documenting the BPF datapath hooks to implement the Cilium datapath, how
+the Cilium datapath integrates with the container orchestration layer, and the
+objects shared between the layers e.g. the BPF datapath and Cilium agent.
+
+Datapath
+============
+
+The Linux kernel supports a set of BPF hooks in the networking stack
+that can be used to run BPF programs. The Cilium datapath uses these
+hooks to load BPF programs that when used together create higher level
+networking constructs.
+
+The following is a list of the hooks used by a Cilium and a brief
+description. For a more thorough documentation on specifics of each
+hook see :ref:`bpf_guide`.
+
+* **XDP:** The XDP BPF hook is at the earliest point possible in the networking driver
+  and triggers a run of the BPF program upon packet reception. This
+  achieves the best possible packet processing performance since the
+  program runs directly on the packet data before any other processing
+  can happen. This hook is ideal for running filtering programs that
+  drop malicious or unexpected traffic and other common DDOS protection
+  mechanisms.
+
+* **Traffic Control Ingress:** BPF programs attached to the traffic control (tc)
+  ingress hook are attached to a networking interface, same as XDP, but
+  will run after the networking stack has done initial processing
+  of the packet. The hook is run before the L3 layer of the stack but has access
+  to most of the metadata associated with a packet. This is ideal for
+  doing local node processing, such as applying L3/L4 endpoint policy
+  and redirecting traffic to endpoints. For networking facing devices the
+  tc ingress hook can be coupled with above XDP hook. When this is done  it
+  is reasonable to assume that the majority of the traffic at this
+  point is legitimate and destined for the host. Containers use a virtual device
+  called a veth pair which acts as a virtual wire connecting the container to
+  the host. By attaching to the TC ingress hook of the host side of this veth pair
+  Cilium can monitor and enforce policy on all traffic exiting a container.
+  By attaching a BPF program to the veth pair from all containers and routing
+  all network traffic to a host side virtual device with another BPF program attached to
+  the tc ingress hook as well Cilium can monitor and enforce policy on all traffic
+  entering/exiting the node.
```

Comment:
Should we also better describe why we need veth with both legs in hostns here as opposed to BPF program on tc and cls_bpf at ingress *and* egress of the physical device?

Reply:
Yes, I think so. IMO I'll look to add that in a separate paragraph though. How about doing that in its own commit.